### PR TITLE
Compute the gradient color from the DOM for NamespaceSelect

### PR DESCRIPTION
**Pull request by Nokel81**
- Once on initial load, and the subsequently only when theme changes
  have applied

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Before: (This is on the beta for v5)
image

### CODE REVIEW

File: src/renderer/utils/color.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * @param field Which computedStyle field to work against
+ */
+export function computeStackingColor(elem: HTMLElement | undefined, field: "color" | "backgroundColor"): Color {
+  if (!elem) {
+    return Color.rgb(0, 0, 0).alpha(0);
+  }
+
+  const curColor = Color(window.getComputedStyle(elem)[field]);
+
+  // console.log(elem);
```
Comment (Nachasic) [inline src/renderer/utils/color.ts]: These probably don't need to be here anymore.

File: src/renderer/utils/color.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Blends the two colors where the parent color is place beneath the child color.
+ * And the returned color is what the browser would render.
+ *
+ * The parent color is placed beneath because child elements are rendered "above"
+ * their DOM parents.
+ * @param parent The color that is from the parent element
+ * @param child The color that is from the child element
+ */
+export function blend(parent: Color, child: Color): Color {
```
Comment (Nachasic) [inline src/renderer/utils/color.ts:43]: Can you provide a comment on what blending mode is implemented here? It's a bit hard to parse the math on it's own.

**Review by Nachasic — COMMENTED** [review]
This looks like a pretty complicated problem to nail down. At the same time, when I inspect the `div.GradientValueContainer` (both under light and dark themes), all I can see in gradient styles is:
```css
element.style {
    right: 0px;
    background: linear-gradient(to left, rgb(255, 255, 255) 0px, transparent);
}
```
which ultimately amounts to no visible gradient at all, since the background color of parent element is white.

**Review by aleksfront — CHANGES_REQUESTED** [review]
We shouldn't overcomplicate things here. I've already had fix, but then it was reverted (since it was part of broken PR).

CSS-only decision must be used for this.

Please use styling provided here
https://github.com/lensapp/lens/pull/2365/commits/4378a58b81be5f1bb1e2312ccc1cdac0cf2da2b5

**Comment by Nokel81** [comment]
@aleksfront should I just use your diff? The problem I see with your fix is that it doesn't work for use in the Wizard.

**Comment by aleksfront** [comment]
> @aleksfront should I just use your diff? The problem I see with your fix is that it doesn't work for use in the Wizard.

I will prepare a separate PR with those fixes asap.

**Comment by Nokel81** [comment]
> This looks like a pretty complicated problem to nail down. At the same time, when I inspect the `div.GradientValueContainer` (both under light and dark themes), all I can see in gradient styles is:
> 
> ```css
> element.style {
>     right: 0px;
>     background: linear-gradient(to left, rgb(255, 255, 255) 0px, transparent);
> }
> ```
> 
> which ultimately amounts to no visible gradient at all, since the background color of parent element is white.

That would be it when used in a `<Wizard>`. But for instance on "Pods" view I see the following:
Screen Shot 2021-05-12 at 8 32 03 AM

**Comment by aleksfront** [comment]
Alternatie fix https://github.com/lensapp/lens/pull/2761

**Comment by Nachasic** [comment]
This is a complicated piece of logic, but it has uses elsewhere and is documented very well, so LGTM

**Review by Nachasic — APPROVED** [review]

File: src/renderer/utils/color.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  const alpha = (1 - child.alpha()) * parent.alpha() + child.alpha();
+  const red = ((1 - child.alpha()) * parent.alpha() * parent.red() + child.alpha() * child.red()) / alpha;
+  const green = ((1 - child.alpha()) * parent.alpha() * parent.green() + child.alpha() * child.green()) / alpha;
+  const blue = ((1 - child.alpha()) * parent.alpha() * parent.blue() + child.alpha() * child.blue()) / alpha;
```
Comment (jim-docker) [inline src/renderer/utils/color.ts:43]: what if alpha is 0?

Comment (Nokel81) [reply src/renderer/utils/color.ts:43]: If alpha is 0 then this algorithm will not take any RGB "value" from that entry because the browser wouldn't either.

**Comment by Nokel81** [comment]
Closing because BG of transparent is now considered a bug, with that being static (wrt the theme) this will no longer be needed.
